### ISSUE TITLE
safe_strcmp and safe_strncmp assuming 8bit return value from strcmp and strncmp

### ISSUE THREAD

**Issue by shufps** [bblanchon/ArduinoJson#1198]
Hi,

the bahaviour of this code is wrong - at least it was on my RISC-V code that is compiled with risc-v g++ 9.2.

The definition of strcmp states that both parameters are equal if the return value is 0 - but the return value is defined as integer and in my case, it tried to compare `addressIndex` with `auth` which gave `0x1100` back. But during the cast to `int8_t` the upper 8bit get lost and the key is accepted as equal. This probably happens during 32bit-wise comparisons that would be totally fine if not only 8 of 32bits are used to determine the end-result.

Probably the same problem also is true for the strncmp variant.

```c++
inline int8_t safe_strcmp(const char* a, const char* b) {
  if (a == b) return 0;
  if (!a) return -1;
  if (!b) return 1;
  return static_cast<int8_t>(strcmp(a, b));
}

inline int8_t safe_strncmp(const char* a, const char* b, size_t n) {
  if (a == b) return 0;
  if (!a) return -1;
  if (!b) return 1;
  return static_cast<int8_t>(strncmp(a, b, n));
}
```

**Comment by bblanchon** [comment]
Hi @shufps,

Thank you very much for reporting this.
This bug was really hard to find.
Luckily, you made a complete diagnostic.
I'll fix that ASAP.

I tend to use 8-bit integers as much as I can because it reduces the size of the code on AVR.
It turned out to be a bad idea here :-(

Best Regards,
Benoit

**Comment by shufps** [comment]
`This bug was really hard to find.`

Haha yes, it took me half the day because I didn't assume the bug could be in the awesome ArduinoJson-lib :sweat_smile: 

Tried to do a minimal replication on my amd64 but the problem didn't appear - so I thought it could be some memory-corruption on my µC-system or side-effects of my memory-protection or ... something else.

So I'm glad, I could find this bug :)

Thank you for fixing!

**Comment by bblanchon** [comment]
This fix was included in [ArduinoJson 6.15.0](https://arduinojson.org/news/2020/03/22/version-6-15-0/).
